task3.py

```diff
@@ -0,0 +1,322 @@
+import os
+import csv
+import requests
+import logging
+import argparse
+from datetime import datetime
+import zipfile
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+file_handler = logging.FileHandler('logs.log')
+file_handler.setLevel(logging.INFO)
+
+formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
+file_handler.setFormatter(formatter)
+
+logger.addHandler(file_handler)
+
+
+def download_csv_file(output_folder, filename):
+    """
+    Downloads a CSV file from a given URL and saves it in the specified output folder
+
+    Parameters:
+        output_folder (str): The path to the output folder where the file will be saved
+        filename (str): The desired filename (without the extension) for the downloaded CSV file
+
+    Returns:
+        str: The path to the downloaded CSV file
+    """
+    url = 'https://randomuser.me/api/?results=5000&format=csv'
+    response = requests.get(url)
+
+    csv_file_path = os.path.join(output_folder, f'{filename}.csv')
+
+    with open(csv_file_path, 'w', encoding='utf-8') as csv_file:
+        csv_file.write(response.text)
+
+    return csv_file_path
+
+
+def filter_data(csv_file, gender, rows):
+    """
+    Filters the data from a CSV file based on gender and a specified number of rows
+
+    Parameters:
+        csv_file (str): The path to the CSV file to be filtered
+        gender (str): The gender to filter by. If None, no gender filtering will be applied
+        rows (int): The maximum number of rows to include in the filtered data. If None, all rows will be included
+
+    Returns:
+        list: A list of dictionaries representing the filtered data
+    """
+    filtered_data = []
+    with open(csv_file, 'r', encoding='utf-8') as csv_file:
+        reader = csv.DictReader(csv_file)
+        for row in reader:
+            if (not gender or row['gender'].lower() == gender.lower()) and (not rows or reader.line_num <= rows):
+                filtered_data.append(row)
+
+    return filtered_data
+
+
+def add_fields(data):
+    """
+    Adds additional fields to the data
+
+    Parameters:
+        data (list): A list of dictionaries representing the data to which additional fields will be added
+    """
+    for row in data:
```

Comment:
try to avoid such **for**.
rewrite function in a waywhere you could apply **map**